### PR TITLE
Prevent multiple upgrades on the same room at once

### PR DESCRIPTION
**Pull request by anoadragon453**
Closes https://github.com/matrix-org/synapse/issues/4583

Does slightly less than https://github.com/matrix-org/synapse/pull/5045, which prevented a room from being upgraded multiple times, one after another. This PR still allows that, but just prevents two from happening at the same time.

Mostly just to mitigate the fact that servers are slow and it can take a moment for the room upgrade to actually complete. We don't want people sending another request to upgrade the room when really they just thought the first didn't go through.

### CODE REVIEW

**Comment by anoadragon453** [comment]
Actually we can probably get rid of the Linearizer now if I'm understanding correctly that it was just supposed to stop two upgrades happening for the same room at once.

**Comment by babolivier** [comment]
> Actually we can probably get rid of the Linearizer now if I'm understanding correctly that it was just supposed to stop two upgrades happening for the same room at once (which this PR now does).

I had a chat with @erikjohnston about this the other day, and the question we might want to ask ourselves is: when two upgrade requests are sent really close in time, what do we want the second to return? If it's an error, then we don't need the Linearizer anymore, but we could instead want to return the same response as the second (as in, when processing the second one, we check whether the room has just been updated, and if so we return the same response as the first upgrade request without going any further in the upgrade process). This solution is better for UX (in that it doesn't prompt an error to the user while the upgrade actually succeeded), and is probably what we want to do. This would require the Linearizer to remain.

**Comment by anoadragon453** [comment]
Ok, but what will be returned if an upgrade succeeds and then another upgrade happens? It should actually complete an upgrade again, correct?

**Comment by babolivier** [comment]
> Ok, but what will be returned if an upgrade succeeds and then another upgrade happens? It should actually complete an upgrade again, correct?

We had a chat about that last night in #synapse-dev, I'm adding the conclusions here for context. We realised it might be interesting to have a behaviour for 2 upgrade requests at the same time for the same room but from two different users different than for these 2 requests at the same time from the same user.

By "at the same time" here I mean that the second request arrives before the upgrade triggered by the first request finishes.

When the two upgrade requests happen from different users, we should return an error for the second request (not sure about the status code though, `409 Conflict` sounds like a good one?) because we think it'd be most likely unintentional for two mods in a room to upgrade the same room at the same time.

When the two upgrade requests happen from the same user, when processing the second request we should skip the upgrade process and return the same response we returned for the first request. This is because such situation is likely unintentional as well (lag/bug/lack of feedback while the upgrade is ongoing), but we don't want to confuse the client (thus the user) in believing the upgrade failed by returning an error while it initially succeeded.

@erikjohnston, we'd like to read your thoughts about this.

File: synapse/handlers/room.py [hunk 1]
```diff
@@ -150,6 +207,15 @@ def upgrade_room(self, requester, old_room_id, new_version):
...
             )
 
+            # Remove the pending request and the entry for this room id if
+            # necessary
+            self.currently_upgrading_rooms[old_room_id]["pending_requests"] -= 1
+            if self.currently_upgrading_rooms[old_room_id]["pending_requests"] <= 0:
+                del self.currently_upgrading_rooms[old_room_id]
+            else:
+                # Save this response if another request is waiting on it
+                self.upgrade_responses[old_room_id] = new_room_id
```
Comment (erikjohnston) [inline synapse/handlers/room.py]: This should be in a `try... finally...`  to properly handle the case of exceptions being thrown.

File: synapse/handlers/room.py [hunk 2]
```diff
@@ -91,6 +100,54 @@ def upgrade_room(self, requester, old_room_id, new_version):
...
+                # homeserver.
+                raise SynapseError(
+                    # 409 - HTTP for "Conflict"
+                    409, "An upgrade for this room is currently in progress",
+                )
+
+        # Mark this room as currently being upgraded by this user
+        self.currently_upgrading_rooms[old_room_id] = {
+            "user": requester.user, "pending_requests": 1,
+        }
```
Comment (erikjohnston) [inline synapse/handlers/room.py]: I think we could probably get rid of a lot of this and reuse `ResponseCache` ?

File: synapse/handlers/room.py [hunk 3]
```diff
@@ -91,66 +100,94 @@ def upgrade_room(self, requester, old_room_id, new_version):
...
-            # start by allocating a new room id
-            r = yield self.store.get_room(old_room_id)
-            if r is None:
-                raise NotFoundError("Unknown room id %s" % (old_room_id,))
-            new_room_id = yield self._generate_room_id(
-                creator_id=user_id, is_public=r["is_public"],
-            )
+        # Check if this room is already being upgraded by another person
+        for key in self._upgrade_response_cache.pending_result_cache:
+            if key[0] == old_room_id and key[1] != user_id:
```
Comment (erikjohnston) [inline synapse/handlers/room.py]: We probably also want to block the same user if  they try to upgrade to different room versions. Currently the response cache below lets the same user upgrade to different versions at the same time.

**Pull request merged**